Bump glutin to 0.29.0

kchibisov opened the pull request:
Fixes #5975.
Fixes #5767.

kchibisov commented:
@sshock could you verify that tab issue you've mentioned in winit https://github.com/rust-windowing/winit/issues/2238 gets resolved with this patch? I'm not sure what to look for myself on macOS for that...

chrisduerr commented on CHANGELOG.md:
```diff
@@ -19,13 +19,19 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - OSC 52 is now disabled on unfocused windows
 - `SpawnNewInstance` no longer inherits initial `--command`
+- Client side decorations should have proper text rendering now on Wayland
+- Search bar is now respecting cursor thickness
```
Not sure this should be fixed in this PR?

kchibisov replied:
It's a part of IME commit.

chrisduerr commented on CHANGELOG.md:
```diff
@@ -37,9 +43,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
+- Incorrect built-in glyphs for `U+2567` and `U+2568`
+- Cursor not hiding on GNOME Wayland
+- Font having different scale factor after monitor powering off/on on X11
+- Wrong mouse input when opening a new tabbed window on macOS
```
Rather than describing the mouse part here, I'd instead mention that the entire window content was offset. It affected rendering too, even with no mouse usage.

chrisduerr commented on alacritty.yml:
```diff
@@ -22,6 +22,20 @@
   # check the local terminfo database and use `alacritty` if it is
   # available, otherwise `xterm-256color` is used.
   #TERM: alacritty
+  #
+  # WINIT_X11_SCALE_FACTOR
+  #
+  # This value is used to control scale factor used by terminal on X11, by
+  # default it automatically detects scale factor of each monitor and apply
+  # it dynamically by using XRandr. Supported values `randr` or floating point
+  # values, like `"1.0"`.
+  #WINIT_X11_SCALE_FACTOR: randr
```
Why did you add this? I really don't want people looking at the config file and thinking it would be a good idea to change this to tweak their scale factor.

kchibisov replied:
I wasn't sure how to put it and how it should be discovered or not. Since we have certain values and I wasn't sure how to translate them.

kchibisov replied:
I think I'll remove env section bits.

chrisduerr commented on alacritty.yml:
```diff
@@ -22,6 +22,20 @@
+  # values, like `"1.0"`.
+  #WINIT_X11_SCALE_FACTOR: randr
+  #
+  # WINIT_WAYLAND_CSD_THEME
+  #
+  # This value is used to pick between Wayland client side decorations variant.
+  # For now only `dark` and `light` are supported.
+  #WINIT_WAYLAND_CSD_THEME: light
```
Should we maybe translate this to an Alacritty option if it's relevant enough to show up in our configuration file?

kchibisov replied:
As an option for config reload? I think we can do that, just not sure that you'd want to.

kchibisov replied:
I've decided to unify it a bit.

chrisduerr commented on alacritty/build.rs:
```diff
@@ -20,7 +20,7 @@ fn main() {
 
     #[cfg(windows)]
-    embed_resource::compile("./windows/windows.rc");
+    embed_resource::compile("./windows/alacritty-icon.rc");
```
Why did you change that here?

kchibisov replied:
The build fails since window-sys creates some lib with `windows` in it's name. So we should have a different name here.

chrisduerr replied:
Then I'd suggest `alacritty`?

kchibisov replied:
I think it'll conflict with the lib from the binary, but we can try...

kchibisov replied:
Seems like it's working.

sshock commented:
> @sshock could you verify that tab issue you've mentioned in winit [rust-windowing/winit#2238](https://github.com/rust-windowing/winit/issues/2238) gets resolved with this patch? I'm not sure what to look for myself on macOS for that...

Yep, it's working!

kchibisov commented on alacritty/src/event.rs:
```diff
@@ -1127,6 +1142,44 @@ impl input::Processor<EventProxy, ActionContext<'_, Notifier, EventProxy>> {
+                        },
+                        // This text is being composed, thus should be shown as inline input to
+                        // the user.
+                        Ime::Preedit(text, cursor_range) => {
+                            // If've got empty preedit we should start processing ordinary keyboard
+                            // input.
+                            if text.is_empty() && cursor_range.is_none() {
+                                self.ctx.display.ime_input.clear_preedit()
+                            } else {
+                                self.ctx.display.ime_input.set_preedit(text, cursor_range);
+                            }
```
What platform are you on?

jayong93 replied:
Oh, I forgot that. I'm on macOS Monterey with M1 Mac.

jayong93 replied:
I've found one more bug. After I insert one or more korean character, ascii characters (like '0' or '+', etc.) are inserted two per key press.

I guess it is because when `ime_input` is enabled, ascii characters are processed both in `WindowEvent::Ime` and `WindowEvent::KeyboardInput`

kchibisov replied:
I've sent a fix upstream, I'll let you know once it merged and get resolved here.

kchibisov commented:
@chrisduerr I've removed IME commit for now and implemented it like we had before. Will rewrite it and send separately. 

Note, that I have bumped msrv to 1.61 for now. Probably will sort that stuff out tomorrow, since it'll require a winit bump to downgrade it back to 1.57.

kchibisov commented on alacritty/Cargo.toml:
```diff
@@ -25,11 +25,11 @@ fnv = "1"
 serde_yaml = "0.8"
 serde_json = "1"
-glutin = { version = "0.28.0", default-features = false, features = ["serde"] }
+glutin = { git = "https://github.com/kchibisov/glutin", branch = "release-029", default-features = false, features = ["serde"] }
```
I'll release it on demand.

kchibisov replied:
glutin was released.

chrisduerr commented on alacritty/src/event.rs:
```diff
@@ -1310,6 +1333,28 @@ impl Processor {
             match event {
+                // The event loop just got initialized. Create a window.
+                GlutinEvent::NewEvents(StartCause::Init) => {
+                    // Creating window inside event loop is required for platforms like macOS to
+                    // properly initialize state, like tab management infra.
+                    //
+                    // Othwerwise the first window won't handle tabs.
```
```suggestion
                    // Creating window inside event loop is required for platforms like macOS to
                    // properly initialize state, like tab management. Othwerwise the first window won't handle tabs.
```

chrisduerr commented on alacritty/src/main.rs:
```diff
@@ -62,6 +61,10 @@ use crate::event::{Event, Processor};
 #[cfg(target_os = "macos")]
 use crate::macos::locale;
 
+// TODO refactor to use `ExitCode` when bumping minimum supported rust version to 1.61.
+pub const EXIT_FAILURE: i32 = 1;
+pub const EXIT_SUCCESS: i32 = 0;
```
I don't think these constants provide any benefit at all in the current code.

kchibisov replied:
See the comment wrt `run_return`.

chrisduerr commented on alacritty/src/main.rs:
```diff
@@ -79,18 +82,23 @@ fn main() {
-        eprintln!("Error: {}", err);
-        process::exit(1);
-    }
+    let exit_code = match result {
+        Ok(code) => code,
+        Err(err) => {
+            eprintln!("Error: {}", err);
+            EXIT_FAILURE
+        },
+    };
```
Why bother with exit codes and all that stuff when `Err(err)` is mapped to an exit code anyway? Just so we don't print the error?

kchibisov replied:
The idea was to propagate the error code from the event loop somehow, since now `run_return` will return non zero code in case of failures.

I basically wanted to indicate an error with non zero exit status nothing more. If you have an idea on how to handle that more reliably let me know.

chrisduerr replied:
Why not return the actual error? That seems to be the sensible choice to me.

chrisduerr reviewed (changes requested):
See my comment about returning a Result.

chrisduerr commented on alacritty/src/main.rs:
```diff
@@ -62,7 +62,7 @@ use crate::event::{Event, Processor};
 use crate::macos::locale;
 
-fn main() {
+fn main() -> Result<(), Box<dyn Error>> {
```
I've only ever gotten extremely poor results from doing this. I don't think it's a great idea?

kchibisov replied:
I think that's fine, like it'll print error? At least that's how I've parsed your lets return `Result`, but now I see that you wanted to return it from `run_return`, but you can't do that.

I'm not sure that this is any different than then `println` and then `exit` that what it should do anyway, no?

chrisduerr replied:
Hm, I think my main issue with last time I tried it is that you don't get any backtrace by doing this. Which should be fine in our case since we were just using `println` anyway.

chrisduerr commented on alacritty/src/event.rs:
```diff
@@ -1310,6 +1337,27 @@ impl Processor {
+                    if let Err(result) =
+                        self.create_window(event_loop, proxy.clone(), initial_window_options)
+                    {
+                        // Stash the error and exit with non-zero code.
+                        *loop_error.lock().unwrap() = result.into();
+                        *control_flow = ControlFlow::ExitWithCode(1);
+                        return;
```
Wait can we not just exit with any generic result, which is then returned from the closure?! That's kinda garbage.

Let's maybe just print the error directly inside the `run_return`, then exit upwards with just any random error text?

I **definitely** don't want to throw mutexes at a silly one-off error that can only happen right before the event loop crashes anyway.

kchibisov replied:
We can sure.

chrisduerr reviewed (approved):
I think it would be nice to have the ability to return from winit's `run_return` with any type parameter, but this should be fine for now.
